rfcs/021 - Datastore Restructuring RFC.md

```diff
@@ -0,0 +1,82 @@
+
+![Datastore arch](./datastore_arch.jpg)
+
+The architecture and flow for the proposed datastore is diagrammed above. The routing layer (whether local mode or remote) is responsible for mutating the HandlerMemory for the given opcode and `opcode.rs` is simply a shell. It chooses the correct node to direct the work to and triggers its counterpart on the other node to request the relevant data from the storage layer, which then eventually returns it. The routing layer hands all of this to the execution layer on the same node to perform the requested work, which then returns it back to the routing layer on that node. The routing layer then stores the mutation, if any, back into the storage layer and then passes the result back to its counterpart on the original node, which then hands the results back to the original datastore opcode in the form of a new HandlerMemory object that replaces the old one (which may simply be the original HandlerMemory object in the case of `mutOnly` calls).
+
+By putting the routing layer as the centerpiece, we can add on greater intelligence to datastore on where it places data in the future. One obvious optimization is to use the DNS regional load balancing that chooses which datacenter to route data to as a "strong hint" on where the primary node should be located to minimize latency for the end user, and we can also go from a fixed replica count to having a replica per datacenter to make eventually consistent reads fast whereever in the world it is located. The AVM gets enough metadata from its DNS-based configuration to do both of these, but in the current structure of the codebase it is very difficult to tackle.
+
+### Alternatives Considered
+
+The primary alternative considered is to just drop `Dashmap` and use a `RWLock`-wrapped `Hashmap` singleton. This would resolve the immediate problem of mutating opcodes not waiting for a proper lock before executing, but would also serialize all datastore accesses on a node, which would be a significant performance regression. It may still be worthwhile to do in the interim if prioritization of this RFC is lower.
```

Comment:
In my opinion if we are going to the "distributed data" path we should tackle this with the proposed architecture right away. Otherwise would be difficult to handle any change.